core/src/main/scala/com/microsoft/azure/synapse/ml/causal/LinearDMLEstimator.scala

```diff
@@ -0,0 +1,335 @@
+          .setModel(getOutcomeModel)
+          .setLabelCol(getOutcomeCol)
+          .setExcludedFeatures(Array(getTreatmentCol)),
+        regressor.getPredictionCol,
+        Seq(regressor.getPredictionCol)
+      )
+    }
+
+    val treatmentModelV1 = treatmentEstimator.fit(train)
+    val outcomeModelV1 = outcomeEstimator.fit(train)
+
+    // Step 3 - use second sample data to get predictions and compute residuals
+    val treatmentPredictedDFV1 = treatmentModelV1.transform(test)
+    val treatmentResidualDFV1 =
+      new ComputeResidualTransformer()
+        .setObservedCol(getTreatmentCol)
+        .setPredictedCol(treatmentResidualPredictionColName)
+        .setOutcomeCol(SchemaConstants.TreatmentResidualColumn)
+        .transform(treatmentPredictedDFV1)
+        .drop(treatmentPredictionColsToDrop: _*)
+
+    val outcomePredictedDFV1 = outcomeModelV1.transform(treatmentResidualDFV1)
+
+    val residualsDFV1 =
+      new ComputeResidualTransformer()
+        .setObservedCol(getOutcomeCol)
+        .setPredictedCol(outcomeResidualPredictionColName)
+        .setOutcomeCol(SchemaConstants.OutcomeResidualColumn)
+        .transform(outcomePredictedDFV1)
+        .drop(outcomePredictionColsToDrop: _*)
+
+    // Step 4 - cross fitting to get another residuals
+    val treatmentModelV2 = treatmentEstimator.fit(test)
+    val outcomeModelV2 = outcomeEstimator.fit(test)
+
+    val treatmentPredictedDFV2 = treatmentModelV2.transform(train)
+    val treatmentResidualDFV2 =
+      new ComputeResidualTransformer()
+        .setObservedCol(getTreatmentCol)
+        .setPredictedCol(treatmentResidualPredictionColName)
+        .setOutcomeCol(SchemaConstants.TreatmentResidualColumn)
+        .transform(treatmentPredictedDFV2)
+        .drop(treatmentPredictionColsToDrop: _*)
+
+    val outcomePredictedDFV2 = outcomeModelV2.transform(treatmentResidualDFV2)
+    val residualsDFV2 =
+      new ComputeResidualTransformer()
+        .setObservedCol(getOutcomeCol)
+        .setPredictedCol(outcomeResidualPredictionColName)
+        .setOutcomeCol(SchemaConstants.OutcomeResidualColumn)
+        .transform(outcomePredictedDFV2)
+        .drop(outcomePredictionColsToDrop: _*)
+
+    // 5. apply regressor fit to get treatment effect T1 = lrm1.coefficients(0) and T2 = lrm2.coefficients(0)
+    val va: Array[PipelineStage] = Array(
+      new VectorAssembler()
+        .setInputCols(Array(SchemaConstants.TreatmentResidualColumn))
+        .setOutputCol("treatmentResidualVec")
+        .setHandleInvalid("skip"),
+      new DropColumns().setCols(Array(SchemaConstants.TreatmentResidualColumn))
+    )
+
+    val regressor = new GeneralizedLinearRegression()
+      .setLabelCol(SchemaConstants.OutcomeResidualColumn)
+      .setFeaturesCol("treatmentResidualVec")
+      .setFamily("gaussian")
+      .setLink("identity")
+      .setFitIntercept(false)
+
+    val lrmPipelineModelV1 = new Pipeline().setStages(va :+ regressor).fit(residualsDFV1)
```

Comment:
Step 3 and 5 should be extracted into a function, and we just call the function twice with two dataframes (train and test) and have them swapped for the second call.

Reply:
after changing to use pipeline, I think code is cleaner now. 
and I think extracting them into a function decrease readability and won't save a lot of duplicated code lines.